cookbooks/custom_formatter.rst

```diff
@@ -0,0 +1,465 @@
+       {
+           return [];
+       }
+
+       /** Forces output to be decorated. */
+       public function setOutputDecorated($decorated): void
+       {
+           $this->isOutputDecorated = (bool) $decorated;
+       }
+
+       /** @deprecated */
+       public function isOutputDecorated()
+       {
+           return $this->isOutputDecorated;
+       }
+
+       /**
+        * Behat can have multiple verbosity levels, you may want to handle this to display more information.
+        * These use the Symfony\Component\Console\Output\OutputInterface::VERBOSITY_ constants.
+        * For reviewdog, I do not need that.
+        */
+       public function setOutputVerbosity($level): void { }
+
+       /** @deprecated */
+       public function getOutputVerbosity()
+       {
+           return 0;
+       }
+
+       /**
+        * Writes message(s) to output stream.
+        *
+        * @param string|array<string> $messages
+        */
+       public function write($messages): void
+       {
+           if (!is_array($messages)) {
+               $messages = [$messages];
+           }
+
+           $this->doWrite($messages, false);
+       }
+
+       /**
+        * Writes newlined message(s) to output stream.
+        *
+        * @param string|array<string> $messages
+        */
+
+       public function writeln($messages = ''): void
+       {
+           if (!is_array($messages)) {
+               $messages = [$messages];
+           }
+
+           $this->doWrite($messages, true);
+       }
+
+       /**
+        * Clear output stream, so on next write formatter will need to init (create) it again.
+        * Not needed in my case.
+        */
+       public function flush(): void
+       {
+       }
+
+       /**
+        * Called by the formatter when test starts
+        */
+       public function removeOldFile(): void
+       {
+           $filePath = $this->getFilePath();
+
+           if (file_exists($filePath)) {
+               unlink($filePath);
+           }
+       }
+
+       /**
+        * @param array<string> $messages
+        */
+       private function doWrite(array $messages, bool $append): void
+       {
+           // create the output path if if does not exists.
+           if (!is_dir($this->getOutputPath())) {
+               mkdir($this->getOutputPath(), 0777, true);
+           }
+
+           // write data to the file
+           file_put_contents($this->getFilePath(), implode("\n", $messages) . "\n", $append ? \FILE_APPEND : 0);
+       }
+
+       private function getFilePath(): string
+       {
+           return $this->getOutputPath() . '/' . $this->fileName;
+       }
+   }
+
+Integration in your project
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You need to add the extension in your Behat configuration file (default is ``behat.yml``) and configure it to use the formatter:
+
+.. code:: yaml
+
+   default:
+     extensions:
+       JDeniau\BehatReviewdogFormatter\ReviewdogFormatterExtension: ~
```

Comment:
if you put your extension class in a `ServiceContainer` subnamespace, with a name matching the last section of your namespace, this unlocks using the shortcut.